### PR TITLE
[8.x] Adds ULID unique identifier support

### PR DESCRIPTION
**Pull request by muhsenmaqsudi**
This PR adds another unique identifier option to the framework beside uuid using `robinvdvleuten/ulid` package

# What is ULID?
ULID (Universally Unique Lexicographically Sortable Identifier) is a case-insensitive 26-character string consisting of 48-bit timestamp (10 characters) and 80-bit randomness (16 characters):
 01AN4Z07BY      79KA1307SR9X4MV3
```
|----------|    |----------------|
 Timestamp          Randomness
   48bits             80bits
```
We can use ULID as identifiers for labeling records while maintaining their time-series order with millisecond resolution. You can use ULID for providing chronologically sortable IDs in distributed systems. ULID doesn’t guarantee the ordering if ULIDs are generated within the same millisecond from multiple machines, but if you have a single coordinator, it’s possible to generate monotonically increasing ULIDs,

## Comparing to Uuid

UUID can be suboptimal for many use-cases because:

* It isn't the most character efficient way of encoding 128 bits of randomness
* UUID v1/v2 is impractical in many environments, as it requires access to a unique, stable MAC address
* UUID v3/v5 requires a unique seed and produces randomly distributed IDs, which can cause fragmentation in many data structures
* UUID v4 provides no other information than randomness which can cause fragmentation in many data structures

Instead, herein is proposed ULID:

```php
Str::ulid() // 01ARZ3NDEKTSV4RRFFQ69G5FAV
```

* 128-bit compatibility with UUID
* 1.21e+24 unique ULIDs per millisecond
* Lexicographically sortable!
* Canonically encoded as a 26 character string, as opposed to the 36 character UUID
* Uses Crockford's base32 for better efficiency and readability (5 bits per character)
* Case insensitive
* No special characters (URL safe)
* Monotonic sort order (correctly detects and handles the same millisecond)

# What this PR adds

* adds a new ulid() method to the Str support class

```php
// Default ulid generator with the current timestamp & uppercase string
Str::ulid(); // 01FDRXQ57VR4K4RASPT9NPAQC0
// Default ulid generator with the current timestamp & lowercase string
Str::ulid(true); // 01fdrxpmg9njp20z3km461fgax
// Default ulid generator with the given datetime & uppercase string
Str::ulid(false, Carbon::now()->subHour()) // 01FDRTD2K8Z664S24X606N14KD
```

* also added all needed methods for creating it's database fields like uuid

```php
Schema::create('foos', function (Blueprint $table) {
    $table->ulid('id')->primary(); // adds primary ulid column 
    $table->foreignUlid('user_id')->constrained()->cascadeOnDelete(); // adds ulid foreignkey
    $table->ulidMorphs('taggable'); // adds ulid morphs
    $table->nullableUlidMorphs('likeable'); // adds nullable ulid morphs
});
```

### CODE REVIEW

**Comment by driesvints** [comment]
Tests seem to be failing so please fix these before marking as ready for review.

**Comment by muhsenmaqsudi** [comment]
@driesvints thanks i will fix them and make it as ready

File: src/Illuminate/Support/Str.php [hunk 1]
```diff
@@ -886,4 +887,18 @@ public static function createUuidsNormally()
...
     }
+
+    /**
+     * Generate a ULID.
+     *
+     * @param  bool  $lowercase
+     * @param  string|null  $dateTime
+     * @return \Ulid\Ulid
+     */
+    public static function ulid(bool $lowercase = false, string $dateTime = null)
```
Comment (dkulyk) [inline src/Illuminate/Support/Str.php:898]: Maybe will be better using DateTimeInterface instead string?

Comment (muhsenmaqsudi) [reply src/Illuminate/Support/Str.php]: i changed that from string to DateTimeInterface

**Comment by GrahamCampbell** [comment]
Why support this, when we could support the much more widely used ksuid concept?

**Comment by muhsenmaqsudi** [comment]
> Why support this, when we could support the much more widely used ksuid concept?

i didn't know about ksuid, i have heard of ulid in javascript world and used it in one of my company recent projects with laravel,
at first we used the uuid after developing for a while with uuid, we started migrating to ulid because we needed a sortable unique identifier that can also generate ids for our older data, so we can sort by our id....

**Comment by muhsenmaqsudi** [comment]
> Why support this, when we could support the much more widely used ksuid concept?

I didn't know about ksuid, i have heard of ulid in javascript world and used it in one of my company's recent project with laravel,
at first we used the uuid, after developing for a while, we started migrating to ulid because we needed a sortable unique identifier that can also generate ids for our older data, so we can sort by our id....
and in the end we can have both of them supported i think laravel really need another identifier alongside uuid that supports timestamps better...

**Comment by taylorotwell** [comment]
I'm not sure we need to add another 3rd party dependency to support this. You can just pull it into your own application and create the columns you need via the migration methods that are already in place.

**Comment by buismaarten** [comment]
@muhsenmaqsudi you could use something like `Str::orderedUuid()`.

See: https://laravel.com/docs/8.x/helpers#method-str-ordered-uuid

**Comment by muhsenmaqsudi** [comment]
@buismaarten we couldn't use that because we already have data that should have a correct ordered id with their created_at date, ordered uuid won't accept past timestamps

**Comment by vanthao03596** [comment]
@muhsenmaqsudi i'm developing package base on your PR [laravel-ulid](https://github.com/vanthao03596/laravel-ulid), hope you can join

**Pull request closed without merging**